framework/src/org/checkerframework/common/signatureinference/SignatureInferenceScenes.java

```diff
@@ -0,0 +1,726 @@
+     * containing the {@link annotations.Annotation}s of an
+     * {@link annotations.el.ATypeElement}.
+     * @param type the {@link annotations.el.ATypeElement}.
+     * @param atf the annotated type factory of a given type system, whose
+     * type hierarchy will be used.
+     * @param tm the underlying type.
+     */
+    private static AnnotatedTypeMirror typeElementToATM(
+            ATypeElement type, AnnotatedTypeFactory atf, TypeMirror tm) {
+        AnnotatedTypeMirror atm = AnnotatedTypeMirror.createType(tm, atf,
+                false);
+        Set<Annotation> annos = getSupportedAnnosInSet(type.tlAnnotationsHere,
+                atf);
+        for (Annotation anno: annos) {
+            AnnotationMirror am = annotationToAnnotationMirror(
+                    anno, atf.getProcessingEnv());
+            atm.addAnnotation(am);
+        }
+        if (tm.getKind() == TypeKind.ARRAY) {
+            AnnotatedArrayType aat = (AnnotatedArrayType) atm;
+            for (ATypeElement innerType : type.innerTypes.values()) {
+                TypeMirror at = aat.getUnderlyingType().getComponentType();
+                aat.setComponentType(typeElementToATM(innerType, atf,at));
+            }
+        }
+        return atm;
+    }
+
+   /**
+    * Updates an {@annotations.el.ATypeElement} to have the annotations of an
+    * {@link org.checkerframework.framework.type.AnnotatedTypeMirror} passed
+    * as argument. Annotations in the original set that are an implementation
+    * detail are ignored and not added to the resulting set. This method also
+    * checks if the {@annotations.el.ATypeElement} has explicit annotations,
+    * and if that is the case no annotations are added.
+    *
+    * @param newATM the AnnotatedTypeMirror whose annotations will be added to
+    * the ATypeElement.
+    * @param oldATM used to check if the ATypeElement has explicit annotations.
+    * @param atf the annotated type factory of a given type system, whose
+    * type hierarchy will be used.
+    * @param typeToUpdate the ATypeElement which will be updated.
+    * @param idx used to write annotations on compound types of an ATypeElement.
+    * @param defLoc the location where the annotation will be added.
+    */
+    private static void updateTypeElementFromATM(AnnotatedTypeMirror newATM,
+            AnnotatedTypeMirror oldATM, AnnotatedTypeFactory atf,
+            ATypeElement typeToUpdate, int idx, DefaultLocation defLoc) {
+        // Clears only the annotations that are supported by atf.
+        // The others stay intact.
+        if (idx == 1) {
+            // This if avoids clearing the annotations multiple times in cases
+            // of type variables and compound types.
+            Set<Annotation> annosToRemove = getSupportedAnnosInSet(
+                    typeToUpdate.tlAnnotationsHere, atf);
+            typeToUpdate.tlAnnotationsHere.removeAll(annosToRemove);
```

Comment:
This is removing annotations from typeToUpdate even if no annotations will be added later. Is this intentional? If so, please mention this behavior in the Javadoc and add a comment here explaining why it's intentional.